xds/src/main/java/io/grpc/xds/XdsClientWrapperForServerSds.java

```diff
@@ -0,0 +1,259 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc.xds;
+
+import com.google.common.annotations.VisibleForTesting;
+import io.envoyproxy.envoy.api.v2.auth.DownstreamTlsContext;
+import io.envoyproxy.envoy.api.v2.core.Node;
+import io.grpc.Internal;
+import io.grpc.Status;
+import io.grpc.SynchronizationContext;
+import io.grpc.internal.ExponentialBackoffPolicy;
+import io.grpc.internal.GrpcUtil;
+import io.grpc.internal.SharedResourceHolder;
+import io.grpc.xds.EnvoyServerProtoData.CidrRange;
+import io.grpc.xds.EnvoyServerProtoData.FilterChain;
+import io.grpc.xds.EnvoyServerProtoData.FilterChainMatch;
+import io.netty.channel.Channel;
+import io.netty.channel.EventLoopGroup;
+import io.netty.channel.epoll.Epoll;
+import io.netty.channel.epoll.EpollEventLoopGroup;
+import io.netty.util.concurrent.DefaultThreadFactory;
+import java.io.IOException;
+import java.net.InetAddress;
+import java.net.InetSocketAddress;
+import java.net.SocketAddress;
+import java.net.UnknownHostException;
+import java.util.List;
+import java.util.NoSuchElementException;
+import java.util.concurrent.ThreadFactory;
+import java.util.concurrent.TimeUnit;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import javax.annotation.Nullable;
+
+/**
+ * Serves as a wrapper for {@link XdsClientImpl} used on the server side by {@link
+ * io.grpc.xds.internal.sds.XdsServerBuilder}.
+ */
+@Internal
+public final class XdsClientWrapperForServerSds {
+  private static final Logger logger =
+      Logger.getLogger(XdsClientWrapperForServerSds.class.getName());
+
+  private static final EventLoopGroupResource eventLoopGroupResource =
+      Epoll.isAvailable() ? new EventLoopGroupResource("GrpcServerXdsClient") : null;
+
+  private EnvoyServerProtoData.Listener curListener;
+  // TODO(sanjaypujare): implement shutting down XdsServer which will need xdsClient reference
+  @SuppressWarnings("unused")
+  @Nullable private XdsClient xdsClient;
+  private final int port;
+
+  /** Creates a new instance. */
+  public static XdsClientWrapperForServerSds newInstance(
+      int port, Bootstrapper bootstrapper, SynchronizationContext syncContext) throws IOException {
+    Bootstrapper.BootstrapInfo bootstrapInfo = bootstrapper.readBootstrap();
+    final List<Bootstrapper.ServerInfo> serverList = bootstrapInfo.getServers();
+    if (serverList.isEmpty()) {
+      throw new NoSuchElementException("No management server provided by bootstrap");
+    }
+    final Node node = bootstrapInfo.getNode();
+    if (eventLoopGroupResource == null) {
+      throw new IllegalStateException(Epoll.unavailabilityCause());
```

Comment:
Good point! If `Epoll` is not available would it be a good idea to use `DefaultEventLoopGroup` ? Or `ScheduledThreadPoolExecutor` ? @voidzcy for the `XdsClientImpl` what will be better?

Reply:
if we can reuse one from `XdsClientImpl` that would be the best. i think `ScheduledThreadPoolExecutor` is better than `DefaultEventLoopGroup`.

Reply:
reuse which one from `XdsClientImpl`? `XdsClientImpl ` doesn't have one of its own and I am passing one to it while creating a `XdsClientImpl` and there is no other `XdsClientImpl` instance to reuse it from unless I am missing something. I will use `ScheduledThreadPoolExecutor`.

Reply:
it was more generic comment. yes please use the `ScheduledThreadPoolExecutor`.

Reply:
I mean `XdsClientImpl` just takes in a `ScheduledExecutorService`, with whatever its concrete implementation is. It doesn't matter from `XdsClientImpl`'s perspective.

Reply:
@voidzcy I think there is some miscommunication here. I know `XdsClientImpl` takes a `ScheduledExecutorService` and I (as the creator of that object) pass a concrete implementation. When @creamsoup mentioned "we can reuse one from XdsClientImpl that would be the best" there isn't anything to reuse so that part threw me off. The plan now is to  use `ScheduledThreadPoolExecutor`. I hope we are all on the same page.

Reply:
Oh  BTW, if @creamsoup meant, do the same thing the current client side user of `XdsClientImpl` does then that is what I did (after tracing laboriously where it came from via XdsNameResolver etc). I don't have the reference handy and I cannot see what it does when epoll is not available.

Reply:
I am following. I thought you were gonna change `XdsClientImpl` and @creamsoup suggested we are able to "reuse" the existing code as is.  So nvm.

Reply:
since it accepts a `ScheduledExecutorService` I never intended to change `XdsClientImpl`. Anyway looks like this is all settled now